File: doc/api-examples/python/monitor-unit.py
#!/usr/bin/python3
#
# Copyright Contributors to the Eclipse BlueChi project
#
# SPDX-License-Identifier: MIT-0

import sys

import dasbus.connection
from dasbus.loop import EventLoop
from dasbus.typing import get_native


def print_dict_changes(old, new):
    for key in sorted(set(old.keys()) | set(new.keys())):
        if key not in old:
            print(f" {key}: {new[key]}")
        elif key in new:
            o = old[key]
            n = new[key]
            if o != n:
                print(f" {key}: {o} -> {n}")


if len(sys.argv) < 2:
    print("No unit name supplied")
    sys.exit(1)

unit_name = sys.argv[1]
node_name = "*"  # Match all

if len(sys.argv) > 2:
    node_name = sys.argv[2]

bus = dasbus.connection.SystemMessageBus()

controller = bus.get_proxy("org.eclipse.bluechi", "/org/eclipse/bluechi")

monitor_path = controller.CreateMonitor()
monitor = bus.get_proxy("org.eclipse.bluechi", monitor_path)

old_values = {}


def unit_property_changed(node, unit, interface, props):
    old_value = old_values.get(node, {})
    new_value = get_native(props)

    print(f"Unit {unit} on node {node} changed for iface {interface}:")
    print_dict_changes(old_value, new_value)

    old_values[node] = {**old_value, **new_value}


def unit_new(node, unit, reason):
    print(f"New Unit {unit} on node {node}, reason: {reason}")


def unit_state_changed(node, unit, active_state, substate, reason):
    print(
        f"Unit {unit} on node {node}, changed to state: {active_state} ({substate}), reason: {reason}"
    )


def unit_removed(node, unit, reason):
    print(f"Removed Unit {unit} on node {node}, reason: {reason}")
    if node in old_values:
        del old_values[node]


monitor.UnitPropertiesChanged.connect(unit_property_changed)
monitor.UnitNew.connect(unit_new)
monitor.UnitStateChanged.connect(unit_state_changed)
monitor.UnitRemoved.connect(unit_removed)

monitor.Subscribe(node_name, unit_name)

if node_name == "":
    print(f"Waiting for changes to unit `{unit_name}` on any node")
else:
    print(f"Waiting for changes to unit `{unit_name}` on node '{node_name}'")

loop = EventLoop()
loop.run()

File: doc/api-examples/rust/monitor-unit.rs
// Copyright Contributors to the Eclipse BlueChi project
//
// SPDX-License-Identifier: MIT-0

use clap::Parser;
use dbus::{arg, blocking::Connection, Message, Path};
use std::time::Duration;

pub struct UnitNewSignal {
    pub node: String,
    pub unit: String,
    pub reason: String,
}

impl arg::AppendAll for UnitNewSignal {
    fn append(&self, i: &mut arg::IterAppend) {
        arg::RefArg::append(&self.node, i);
    }
}

impl arg::ReadAll for UnitNewSignal {
    fn read(i: &mut arg::Iter) -> Result<Self, arg::TypeMismatchError> {
        Ok(UnitNewSignal {
            node: i.read()?,
            unit: i.read()?,
            reason: i.read()?,
        })
    }
}

impl dbus::message::SignalArgs for UnitNewSignal {
    const NAME: &'static str = "UnitNew";
    const INTERFACE: &'static str = "org.eclipse.bluechi.Monitor";
}

pub struct UnitRemovedSignal {
    pub node: String,
    pub unit: String,
    pub reason: String,
}

impl arg::AppendAll for UnitRemovedSignal {
    fn append(&self, i: &mut arg::IterAppend) {
        arg::RefArg::append(&self.node, i);
    }
}

impl arg::ReadAll for UnitRemovedSignal {
    fn read(i: &mut arg::Iter) -> Result<Self, arg::TypeMismatchError> {
        Ok(UnitRemovedSignal {
            node: i.read()?,
            unit: i.read()?,
            reason: i.read()?,
        })
    }
}

impl dbus::message::SignalArgs for UnitRemovedSignal {
    const NAME: &'static str = "UnitRemoved";
    const INTERFACE: &'static str = "org.eclipse.bluechi.Monitor";
}

pub struct UnitPropertiesChangedSignal {
    pub node: String,
    pub unit: String,
    pub interface: String,
}

impl arg::AppendAll for UnitPropertiesChangedSignal {
    fn append(&self, i: &mut arg::IterAppend) {
        arg::RefArg::append(&self.node, i);
    }
}

impl arg::ReadAll for UnitPropertiesChangedSignal {
    fn read(i: &mut arg::Iter) -> Result<Self, arg::TypeMismatchError> {
        Ok(UnitPropertiesChangedSignal {
            node: i.read()?,
            unit: i.read()?,
            interface: i.read()?,
        })
    }
}

impl dbus::message::SignalArgs for UnitPropertiesChangedSignal {
    const NAME: &'static str = "UnitPropertiesChanged";
    const INTERFACE: &'static str = "org.eclipse.bluechi.Monitor";
}

pub struct UnitStateChangedSignal {
    pub node: String,
    pub unit: String,
    pub active_state: String,
    pub sub_state: String,
    pub reason: String,
}

impl arg::AppendAll for UnitStateChangedSignal {
    fn append(&self, i: &mut arg::IterAppend) {
        arg::RefArg::append(&self.node, i);
    }
}

impl arg::ReadAll for UnitStateChangedSignal {
    fn read(i: &mut arg::Iter) -> Result<Self, arg::TypeMismatchError> {
        Ok(UnitStateChangedSignal {
            node: i.read()?,
            unit: i.read()?,
            active_state: i.read()?,
            sub_state: i.read()?,
            reason: i.read()?,
        })
    }
}

impl dbus::message::SignalArgs for UnitStateChangedSignal {
    const NAME: &'static str = "UnitStateChanged";
    const INTERFACE: &'static str = "org.eclipse.bluechi.Monitor";
}

#[derive(Parser)]
struct Cli {
    /// The node name to list the units for
    #[clap(short, long)]
    unit_name: String,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Cli::parse();
    let unit_name = args.unit_name;
    let node_name = "*"; // Match all nodes

    let conn = Connection::new_system()?;

    let bluechi = conn.with_proxy(
        "org.eclipse.bluechi",
        "/org/eclipse/bluechi",
        Duration::from_millis(5000),
    );

    let (monitor,): (Path,) =
        bluechi.method_call("org.eclipse.bluechi.Controller", "CreateMonitor", ())?;
    let monitor_proxy =
        conn.with_proxy("org.eclipse.bluechi", monitor, Duration::from_millis(5000));

    let _id =
        monitor_proxy.match_signal(move |signal: UnitNewSignal, _: &Connection, _: &Message| {
            println!(
                "New Unit {} on node {}, reason: {}",
                signal.unit, signal.node, signal.reason
            );
            true
        });

    let _id = monitor_proxy.match_signal(
        move |signal: UnitRemovedSignal, _: &Connection, _: &Message| {
            println!(
                "Removed Unit {} on node {}, reason: {}",
                signal.unit, signal.node, signal.reason
            );
            true
        },
    );

    let _id = monitor_proxy.match_signal(
        move |signal: UnitPropertiesChangedSignal, _: &Connection, _: &Message| {
            println!(
                "Unit {} on node {} changed for iface {}",
                signal.unit, signal.node, signal.interface
            );
            true
        },
    );

    let _id = monitor_proxy.match_signal(
        move |signal: UnitStateChangedSignal, _: &Connection, _: &Message| {
            println!(
                "Unit {} on node {} changed to state({}, {}), reason: {}",
                signal.unit, signal.node, signal.active_state, signal.sub_state, signal.reason
            );
            true
        },
    );

    monitor_proxy.method_call(
        "org.eclipse.bluechi.Monitor",
        "Subscribe",
        (node_name, unit_name),
    )?;

    loop {
        conn.process(Duration::from_millis(1000))?;
    }
}
